### PR TITLE
limit number of concurrent rpc calls

### PR DESCRIPTION
**Pull request by segfaultdoc**
can overload system when spawning too many concurrent network calls

### CODE REVIEW

File: tip-distributor/src/lib.rs [hunk 1]
```diff
@@ -513,12 +521,17 @@ pub async fn sign_and_send_transactions_with_retries(
                     }
                     Err(e) => {
                         error!("error sending transaction {sig:?} error: {e:?}");
-                        Err(e)
+                        if let ErrorKind::TransactionError(AlreadyProcessed) = e.kind {
+                            Ok(())
+                        } else {
+                            Err(e)
```
Comment (esemeniuc) [inline tip-distributor/src/lib.rs:527]: maybe move line 523 to be inside the block so we aren't spammed with already processed msgs

**Review by esemeniuc — APPROVED** [review]

File: tip-distributor/src/lib.rs [hunk 2]
```diff
@@ -494,14 +499,17 @@ pub async fn sign_and_send_transactions_with_retries(
...
-            .map(|(sig, tx)| async move {
+        let futs = signatures_to_transactions.iter().map(|(sig, tx)| {
+            let semaphore = semaphore.clone();
+            async move {
+                let permit = semaphore.clone().acquire_owned().await.unwrap();
                 let res = match rpc_client.send_transaction(tx).await {
                     Ok(_sig) => {
                         info!("sent transaction: {_sig:?}");
+                        drop(permit);
                         sleep(Duration::from_secs(5)).await;
```
[the remark this answers is not in the document]
Comment (segfaultdoc) [reply tip-distributor/src/lib.rs:510]: the sleep is so we give the tx enough time to confirm

File: tip-distributor/src/lib.rs [hunk 3]
```diff
@@ -494,14 +499,17 @@ pub async fn sign_and_send_transactions_with_retries(
                 });
         }
 
-        let futs = signatures_to_transactions
-            .iter()
-            .map(|(sig, tx)| async move {
+        let futs = signatures_to_transactions.iter().map(|(sig, tx)| {
```
Comment (buffalu) [inline tip-distributor/src/lib.rs:502]: wb single threading sends here (sends are over localhost so super fast) then single threading confirmations (same wrt speed)

File: tip-distributor/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (segfaultdoc) [reply tip-distributor/src/lib.rs:510]: the sleep is so we give the tx enough time to confirm

File: tip-distributor/src/lib.rs (hunk 3 above)
Comment (buffalu) [reply tip-distributor/src/lib.rs:502]: alternatively, could limit number of tokio worker threads

Comment (buffalu) [reply tip-distributor/src/lib.rs:502]: and not use semaphore

Comment (esemeniuc) [reply tip-distributor/src/lib.rs:502]: could do:
```
let mut rt = runtime::Builder::new()
    .core_threads(4)
    .build()
    .unwrap();

```

**Pull request merged**